packages/proxy-preview/src/index.ts

```diff
@@ -0,0 +1,270 @@
+import cookie from "cookie";
+import { Toucan } from "toucan-js";
+
+export interface Env {
+	SENTRY_DSN: string;
+}
+
+class HttpError extends Error {
+	constructor(message: string, readonly status: number) {
+		super(message);
+	}
+	toResponse() {
+		return Response.json(
+			{
+				error: this.name,
+				message: this.message,
+			},
+			{ status: this.status }
+		);
+	}
+}
+
+class NoExchangeUrl extends HttpError {}
+class ExchangeFailed extends HttpError {}
+class TokenUpdateFailed extends HttpError {}
+class RawHttpFailed extends HttpError {}
+class PreviewRequestFailed extends HttpError {}
+
+// No ecosystem routers support hostname matching 😥
+export default {
+	async fetch(
+		request: Request,
+		env: Env,
+		ctx: ExecutionContext
+	): Promise<Response> {
+		const sentry = new Toucan({
+			dsn: env.SENTRY_DSN,
+			context: ctx,
+			request,
+			requestDataOptions: {
+				allowedHeaders: [
+					"user-agent",
+					"accept-encoding",
+					"accept-language",
+					"cf-ray",
+					"content-length",
+					"content-type",
+					"host",
+				],
+			},
+		});
+		try {
+			const url = new URL(request.url);
+
+			/**
+			 * Request the preview session associated with a given exchange_url
+			 * exchange_url comes from an authenticated core API call made in the client
+			 * It doesn't have CORs set up, so needs to be proxied
+			 */
+			if (
+				request.method === "POST" &&
+				url.hostname === "preview.devprod.cloudflare.dev" &&
+				url.pathname === "/exchange"
+			) {
+				const exchangeUrl = url.searchParams.get("exchange_url");
+				if (!exchangeUrl) {
+					throw new NoExchangeUrl("No exchange_url provided", 400);
+				}
+				const exchangeRes = await fetch(exchangeUrl);
+				if (exchangeRes.status !== 200) {
+					const exchange = new URL(exchangeUrl);
+					// Clear sensitive token
+					exchange.search = "";
+					console.error(
+						"Failed Exchange",
+						exchange.href,
+						exchangeRes.status,
+						await exchangeRes.text()
+					);
+					throw new ExchangeFailed("Exchange failed", 400);
+				}
+				const session = (await exchangeRes.json()) as {
+					prewarm: string;
+					token: string;
+				};
+				if (
+					typeof session.token !== "string" ||
+					typeof session.prewarm !== "string"
+				) {
+					const exchange = new URL(exchangeUrl);
+					// Clear sensitive token
+					exchange.search = "";
```

Comment:
Do we care about doing this?
The console.error will only be visible if you tail this worker, and not to general public. Right?

Reply:
I'm not sure about this, since it's going to sentry—I feel like we should strip out tokens?